src/validation-engine.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  ValidationResult,
+  FormValidationResult,
+  FieldsValidationSchema,
+  RecordValidationFunction,
+  createDefaultValidationResult,
+  RecordValidationFunctionSyncAsync,
+  isSyncValidationResult,
+  FieldValidationFunctionSyncAsync,
+  FieldValidationFunction,
+  FieldValidation,
+} from './model';
+
+import { isUndefinedOrNull } from './helper';
+
+import {
+  fireAllFieldsValidations,
+  fireSingleFieldValidations,
+  fireRecordValidations,
+} from './validation-dispatcher';
+import { buildFormValidationResult } from './form-validation-summary-builder';
+
+export class ValidationEngine {
+  private validationsPerField: FieldsValidationSchema = {};
+  private validationsGlobalForm: RecordValidationFunction[] = [];
+  private asyncValidationInProgressCount = 0;
+
+  // TODO, in new api we are adding an optional parameter
+  // to customize the error message (string array)
+  //
+  // TODO: Chain pattern applied here, check why it is useful
+  addFieldValidation(
+    key: string,
+    validation: FieldValidationFunctionSyncAsync,
+    customParams: any = {},
+    errorMessage?: string
+  ): ValidationEngine {
+    const asyncValidationFn = this.convertFieldValidationToAsyncIfNeeded(
+      validation
+    );
+
+    if (!this.isFieldKeyMappingDefined(key)) {
+      this.validationsPerField[key] = [];
+    }
+
+    const fieldValidation: FieldValidation = {
```

Comment:
We are not using this const, isn't it?

Reply:
yes, in this line:

this.validationsPerField[key].push(fieldValidation);